### ISSUE TITLE
EXIF Orientation not supported

### ISSUE THREAD

**Issue by Yay295** [shssoichiro/oxipng#570]
I created two images with ImageMagick and added the EXIF Orientation tag to both of them:
```
convert -size 64x128 canvas:#FFFF00 exif_rotated.png
%EXIFTOOL% -overwrite_original -n -Orientation=6 exif_rotated.png

convert -size 64x128 -delay 100 canvas:#FFFF00 canvas:#0000FF exif_rotated_a.png
%EXIFTOOL% -overwrite_original -n -Orientation=6 exif_rotated_a.png
```
![exif_rotated](https://github.com/shssoichiro/oxipng/assets/8922038/a5c3c868-e5ca-4994-b095-d150dd70b165)
![exif_rotated_a](https://github.com/shssoichiro/oxipng/assets/8922038/1a0242f9-0a2d-4a46-8790-f3c82531e3a1)
After compressing them with oxipng, the EXIF tags have been stripped, and the images have not been rotated. Note that neither Firefox nor Chrome currently respect this tag, and Safari only sometimes does, so these images will not be displayed correctly in your browser.
https://bugzilla.mozilla.org/show_bug.cgi?id=1627423
https://bugs.webkit.org/show_bug.cgi?id=210021

**Comment by andrews05** [comment]
Hi @Yay295 
Oxipng will not strip anything by default - this will only happen if you are using the `--strip` flag. If you don't want the EXIF tags to be stripped you'll need to either not use `--strip` or be more explicit about what chunks you want to strip/keep.

**Comment by Yay295** [comment]
I was using `--strip safe`. I don't think EXIF orientation is a safe tag to strip.

**Comment by andrews05** [comment]
EXIF is an interesting situation. It's only recently been adopted into the PNG spec (third edition, still not a final standard yet) and as you pointed out is not widely supported yet. EXIF can also contain a huge variety of different metadata and much of the time it is unimportant and perfectly safe to strip. Without adding full EXIF support to oxipng to read and understand the data, we can't know whether it might be "safe" or not. It's certainly a feature we could consider, but it's not a simple matter.

**Comment by AlexTMjugador** [comment]
> Without adding full EXIF support to oxipng to read and understand the data, we can't know whether it might be "safe" or not. It's certainly a feature we could consider, but it's not a simple matter.

There's the [`exif`](https://docs.rs/kamadak-exif/0.5.5/exif/) crate for parsing EXIF tags which happens to mention EXIF orientation tags in its examples, and has only a single dependency. Its source package weighs 53.2 KiB, though, which makes me wonder "do we really want to include this rather large dependency to solve this use case?". But the building blocks are there, and it should not be too hard to do some research on which EXIF tags are not safe to strip.

**Comment by ace-dent** [comment]
Personal opinion:
Any metadata that could have been 'baked' into the raw RGBA, but wasn't, can be stripped as part of 'lossless' optimization. 
i.e. removing gamma, color transforms, and now rotation. 
If you want to preserve metadata, don't use `--strip`.
Or use a script to perform the transformation and then strip the metadata.

**Comment by Yay295** [comment]
If you want to remove all metadata you can use `--strip all`. `--strip safe` exists specifically to *not* strip metadata like this though.

**Comment by Yay295** [comment]
Theoretically, you could actually save space by adding an orientation tag if you were able to compress the image better by flipping/rotating it.

**Comment by andrews05** [comment]
I think we would want more data on this before we can make a call on it. How and why EXIF is being used in pngs, what tags etc. Your report is helpful, but currently it’s the only data point we have.

Right now I don’t think it’s a big issue. If you need to preserve it, you can. If you strip it inadvertently, it’s easy enough to restore (unlike some other chunks).

**Comment by Yay295** [comment]
Yeah, it's not a problem for me. I only noticed it because I added code to my scripts to check for it, so I'm already handling it. I just think it's something that oxipng should check for.

**Comment by kjkent** [comment]
So I'm in the process of optimizing my image library, but want to preserve things like EXIF tags so that images appear at the correct points in visual timelines. `--strip safe` would be a useful option for me if it left these tags (or at least specific tags such as timestamps and geolocation) while stripping other "safe" data.

As `--strip safe` _does_ remove these tags, and I don't know enough about all other types of metadata chunks to list them under `--keep`, I'll run oxipng without stripping them.

This comment is just intended to give an example of a use case where this functionality would be helpful. Thanks!

**Comment by andrews05** [comment]
@kjkent Since this issue was opened, we've added a new way to use safe stripping while keeping additional chunks. In this case you would use `--keep display,eXIf`, where the special value `display` refers to all the same chunks that `--strip safe` keeps.

**Comment by kjkent** [comment]
Amazing, that's great to know, thank you! The help string for the CLI might benefit from pointing that last bit out; would you accept a small PR?

**Comment by andrews05** [comment]
It is mentioned in the help, is there some way you feel it could be improved?
[edit] You do need to use the long `--help` to see this, rather than the short `-h`.

**Comment by kjkent** [comment]
@andrews05 apologies for my delayed response here, and right you are. I must have totally missed that. Thank you, in any case :)

**Comment by Yay295** [comment]
Both Firefox and Chrome now support this tag, but it doesn't look like anything has changed for WebKit.

**Comment by andrews05** [comment]
Thanks for the update, it's looking like we have good reason to do something about this.
What I'm thinking is:
- No change to stripping rules.
- If eXIf is not stripped (i.e. no strip option specified, or by explicitly keeping it) then no action required.
- If eXIf is stripped (i.e. with strip safe) then rewrite the image to match the orientation tag that was lost. Similar to exifautotran.

**Comment by TPS** [comment]
> Theoretically, you could actually save space by adding an orientation tag if you were able to compress the image better by flipping/rotating it.

I actually have [an (extreme) example pair](https://github.com/ace-dent/columbo/issues/1#issuecomment-4953583291) for this.  Is it worth pursuing as a potential optimization technique, than taking only the safe path of handling EXIF properly?

**Comment by andrews05** [comment]
Users can certainly explore such things themselves, but it's not appropriate for oxipng to be doing that.
